plugins/inputs/statsd/datadog.go

```diff
@@ -0,0 +1,173 @@
+func (s *Statsd) parseEventMessage(now time.Time, message string, defaultHostname string) error {
+	// _e{title.length,text.length}:title|text
+	//  [
+	//   |d:date_happened
+	//   |p:priority
+	//   |h:hostname
+	//   |t:alert_type
+	//   |s:source_type_nam
+	//   |#tag1,tag2
+	//  ]
+	//
+	//
+	// tag is key:value
+	messageRaw := strings.SplitN(message, ":", 2)
+	if len(messageRaw) < 2 || len(messageRaw[0]) < 7 || len(messageRaw[1]) < 3 {
+		return fmt.Errorf("Invalid message format")
+	}
+	header := messageRaw[0]
+	message = messageRaw[1]
+
+	rawLen := strings.SplitN(header[3:], ",", 2)
+	if len(rawLen) != 2 {
+		return fmt.Errorf("Invalid message format")
+	}
+
+	titleLen, err := strconv.ParseInt(rawLen[0], 10, 64)
+	if err != nil {
+		return fmt.Errorf("Invalid message format, could not parse title.length: '%s'", rawLen[0])
+	}
+
+	textLen, err := strconv.ParseInt(rawLen[1][:len(rawLen[1])-1], 10, 64)
+	if err != nil {
+		return fmt.Errorf("Invalid message format, could not parse text.length: '%s'", rawLen[0])
+	}
+	if titleLen+textLen+1 > int64(len(message)) {
+		return fmt.Errorf("Invalid message format, title.length and text.length exceed total message length")
+	}
+
+	rawTitle := message[:titleLen]
+	rawText := message[titleLen+1 : titleLen+1+textLen]
+	message = message[titleLen+1+textLen:]
+
+	if len(rawTitle) == 0 || len(rawText) == 0 {
+		return fmt.Errorf("Invalid event message format: empty 'title' or 'text' field")
+	}
+
+	// Handle hostname, with a priority to the h: field, then the host:
+	// tag and finally the defaultHostname value
+	// Metadata
+	m := cachedEvent{
+		name: rawTitle,
+	}
+	m.tags = make(map[string]string, strings.Count(message, ",")+2) // allocate for the approximate number of tags
+	m.fields = make(map[string]interface{}, 9)
+	m.fields["alert-type"] = "info" // default event type
+	m.fields["text"] = uncommenter.Replace(string(rawText))
+	m.tags["source"] = defaultHostname
+	m.fields["priority"] = priorityNormal
+	m.ts = now
+	if len(message) == 0 {
+		s.events = append(s.events, m)
+		return nil
+	}
+
+	if len(message) > 1 {
+		rawMetadataFields := strings.Split(message[1:], "|")
+		for i := range rawMetadataFields {
+			if len(rawMetadataFields[i]) < 2 {
+				log.Printf("W! [inputs.statsd] too short metadata field")
+			}
+			switch rawMetadataFields[i][:2] {
+			case "d:":
+				ts, err := strconv.ParseInt(rawMetadataFields[i][2:], 10, 64)
+				if err != nil {
+					log.Printf("W! [inputs.statsd] skipping timestamp: %s", err)
+					continue
+				}
+				m.fields["ts"] = ts
+			case "p:":
+				switch rawMetadataFields[i][2:] {
+				case priorityLow:
+					m.fields["priority"] = priorityLow
+				case priorityNormal: // we already used this as a default
+				default:
+					log.Printf("W! [inputs.statsd] skipping priority")
+					continue
+				}
+			case "h:":
+				m.tags["source"] = rawMetadataFields[i][2:]
```

Comment:
@danielnelson What's typical for a source host in telegraf?

@docmerlin was "source" used in other places too?

Reply:
[This is the discussion](https://github.com/influxdata/telegraf/issues/4413) on standardizing what `source` should be set to when used as a tag.